ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,819 @@
+	pub fn add_notify(&self, target: Arc<ChainNotify>) {
+		self.notify.write().push(Arc::downgrade(&target));
+	}
+
+	fn notify<F>(&self, f: F) where F: Fn(&ChainNotify) {
+		for np in self.notify.read().iter() {
+			if let Some(n) = np.upgrade() {
+				f(&*n);
+			}
+		}
+	}
+
+	/// Register client reference
+	pub fn register_client(&self, client: Weak<Client>) {
+		*self.client.write() = Some(client);
+	}
+
+	/// Register accounts provider
+	pub fn register_account_provider(&self, accounts: Weak<AccountProvider>) {
+		*self.accounts.write() = Some(accounts);
+	}
+
+	/// Sets provider's config. Required mostly in tests
+	pub fn set_config(&self, config: ProviderConfig) {
+		*self.config.lock() = config;
+	}
+
+	/// 1. Create private transaction from the regular transaction
+	/// 2. Executes private transaction
+	/// 3. Save it with state returned on prev step to the queue for signing
+	/// 4. Broadcast corresponding message to the chain
+	pub fn create_private_transaction(&self, transaction: SignedTransaction, contract: &Address) -> Result<(), EthcoreError> {
+		trace!("Creating private transaction from regular transaction: {:?}", transaction);
+		let data = transaction.rlp_bytes();
+		let encrypted_transaction = self.encrypt(contract, &data)?;
+		let private = PrivateTransaction {
+			encrypted: encrypted_transaction,
+			contract: *contract,
+		};
+		let private_state = self.execute_private_transaction(BlockId::Latest, &transaction)?;
+		trace!("Private transaction created, encrypted transaction: {:?}, private state: {:?}", private, private_state);
+		let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+		trace!("Required validators: {:?}", contract_validators);
+		self.transactions_for_signing.lock().add_transaction(private.hash(), transaction, contract_validators, private_state)?;
+		self.broadcast_private_transaction(private.rlp_bytes().into_vec());
+		Ok(())
+	}
+
+	/// Extract signed transaction from private transaction
+	fn extract_original_transaction(&self, private: PrivateTransaction, contract: &Address) -> Result<UnverifiedTransaction, EthcoreError> {
+		let encrypted_transaction = private.encrypted.clone();
+		let transaction_bytes = self.decrypt(contract, &encrypted_transaction)?;
+		let original_transaction: UnverifiedTransaction = UntrustedRlp::new(&transaction_bytes).as_val()?;
+		Ok(original_transaction)
+	}
+
+	/// Process received private transaction
+	pub fn import_private_transaction(&self, rlp: &[u8]) -> Result<(), EthcoreError> {
+		let validator_accounts = self.config.lock().validator_accounts.clone();
+		trace!("Private transaction received");
+		if validator_accounts.is_empty() {
+			self.broadcast_private_transaction(rlp.into());
+			return Ok(());
+		}
+		let private_tx: PrivateTransaction = UntrustedRlp::new(rlp).as_val()?;
+		let contract = private_tx.contract;
+		let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+		let addresses_intersection = contract_validators
```

Comment:
`iter().find(<predicate>).is_some()` can be written more efficiently: `iter().any(<predicate>)`.

Also, you don't need `filter` and `collect` here, get an `Option` for the first element -- and just check if it is Some.